woocommerce/woocommerce-ios
# [Mobile Payments] Fix card reader connection modal transitions

Closes: #7686 


### Description


When connecting a card reader, and choosing between multiple readers, some of the transitions weren't quite right.

This PR changes the flow to use a system transition style for all modal presentation and dismissal, and corrects the `animated` flags for each transition. It doesn't resolve all the issues identified, but gets most of them without making major changes.

- [x] Connection flow (`Scanning for reader`) does not animate in.
- [ ] There is a brief white flash caused by the overlay view disappearing between `Keep searching` and the `Several readers found` modal.
- [x] The grey overlay view is a different colour between the connection modal and the `Several readers found` modal.
- [x] After tapping `Connect` on a reader, the `Several readers found` modal is animated with a slide down off screen, then immediately replaced with the `Connecting to reader` modal.
- [x] For some of the time `Connecting to reader` is on screen, an extra overlay is placed over the whole screen.
- [x] When the reader connection is complete, the grey overlay is removed before the modal above it.
- [x] The modal is not animated off at the end of the flow.

The design of the modal flow seems to prevent us from fixing the brief white flash when the overlay view disappears. Switching between modals is two operations: first dismiss, then present, and the overlay is part of the views that are dismissed and presented. A redesign could have a host view controller be responsible for the overlay, then the display and dismissal of the various modals, only dismissing the host controller when the flow ends.

The animated dismissal and replacement of the `Several readers found` screen is also a brief white flash now.

> - [x] After tapping `Connect` on a reader, the `Several readers found` modal is animated with a slide down off screen, then immediately replaced with the `Connecting to reader` modal.

### Testing instructions


You need multiple readers to test this; the simulated readers can also be used if you run it from Xcode.
Use `Edit Scheme > Arguments > -simulate-stripe-card-reader ✅` to use the simulator.

1. Turn on all your card readers
2. Navigate to `Menu > Payments`
3. Tap `Manage card reader`
4. Tap `Connect card reader`
5. Tap `Keep searching` when the first reader comes up*
6. Tap connect on a reader in the list of readers, when it shows

Repeat the above and similar steps, including cancel at different steps. Check that the transitions are more consistent than in the Before video below (they aren't perfect, just better.)

The flow can also be accessed when taking a payment without a connected reader: see `Payments > Collect payment`, `Orders > + > Create Order`, and `Orders > Order details > Collect Payment` for these flows.

*This will only come up if you've not got a saved reader which it auto-connects to. If it auto-connects, just tap `Disconnect` for that reader and try again.

### Screenshots


#### Before 
https://user-images.githubusercontent.com/2472348/189184113-429f1f14-5a44-4b9f-8da4-572e2f7ed6d9.mp4

#### After
https://user-images.githubusercontent.com/2472348/189184595-ef94ce0c-ec6e-48e4-8949-6ebbbbf7174d.mp4

---
